### DISCUSSION TITLE
I got an error, please help.

### DISCUSSION THREAD

**Discussion by Godwhitelight** [spotify-web-api-java/spotify-web-api-java#261] [Q&A] [answered]
java.lang.NoClassDefFoundError: org/apache/hc/client5/http/auth/Credentials
	at com.wrapper.spotify.SpotifyHttpManager$Builder.build(SpotifyHttpManager.java:392)
	at com.wrapper.spotify.SpotifyApi.<clinit>(SpotifyApi.java:73)
	at com.wrapper.spotify.SpotifyApi$Builder.<init>(SpotifyApi.java:2012)
	at God.Shush.GuiInGameHook.<clinit>(GuiInGameHook.java:65)
	at net.minecraft.client.Minecraft.startGame(Minecraft.java:565)
	at net.minecraft.client.Minecraft.run(Minecraft.java:402)
	at net.minecraft.client.main.Main.main(Main.java:113)
	at Start.main(Start.java:7)
Caused by: java.lang.ClassNotFoundException: org.apache.hc.client5.http.auth.Credentials
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	... 8 more
	
I think it says that its got invalid credentials. but I have done it like the example of the get the current track and it's not working. so I got an OAuth and checked it from the Spotify website. and it's working there. but on here its won't.

**Comment by KoolieAid** [comment]
Do you have a shade / shadow plugin in your build tool?

**Comment by dargmuesli** [comment]
> I think it says that its got invalid credentials

It says `java.lang.NoClassDefFoundError`... your project cannot find a class, it doesn't say that authentication failed. You'd have to share your project's source code for us to be able to help you, otherwise we can only guess.

Also, please post questions to Stackoverflow and not to this issue tracker: https://github.com/spotify-web-api-java/spotify-web-api-java/blob/develop/CONTRIBUTING.md#your-questions

**Comment by dargmuesli** [comment]
Or, alternatively, to GitHub discussions, which I have just now enabled for this repository.

**Comment by Godwhitelight** [comment, marked as answer]
```
private static final String clientId = "<redacted>";
    private static final String clientSecret = "<redacted>";

    private static final SpotifyApi spotifyApi = new SpotifyApi.Builder()
            .setClientId(clientId)
            .setClientSecret(clientSecret)
            .build();
    private static final ClientCredentialsRequest clientCredentialsRequest = spotifyApi.clientCredentials()
            .build();

    public static void clientCredentials_Sync() {
        try {
            final ClientCredentials clientCredentials = clientCredentialsRequest.execute();

            // Set access token for further "spotifyApi" object usage
            spotifyApi.setAccessToken(clientCredentials.getAccessToken());

            System.out.println("Expires in: " + clientCredentials.getExpiresIn());
        } catch (IOException | SpotifyWebApiException | ParseException e) {
            System.out.println("Error: " + e.getMessage());
        }
    }
    private static final GetUsersCurrentlyPlayingTrackRequest getUsersCurrentlyPlayingTrackRequest = spotifyApi
            .getUsersCurrentlyPlayingTrack()
//          .market(CountryCode.SE)
//          .additionalTypes("track,episode")
            .build();

    public static void getUsersCurrentlyPlayingTrack_Sync() {
        try {
            final CurrentlyPlaying currentlyPlaying = getUsersCurrentlyPlayingTrackRequest.execute();

            System.out.println("Timestamp: " + currentlyPlaying.getTimestamp());
        } catch (IOException | SpotifyWebApiException | ParseException e) {
            System.out.println("Error: " + e.getMessage());
        }
    }


    private void renderSpotify() {
        getUsersCurrentlyPlayingTrack_Sync();
    }
```
here is my code then

**Reply by dargmuesli** [reply]
Never post your client secret publicly! Consider it leaked now. Eventhough I redacted it from your post, my change can still be seen in the changelog.

Again, we cannot help until you share your **project's** source code. Not just the content of a single file. This file cannot explain why your project's setup might not be able to find a dependency class.

**Reply by Godwhitelight** [reply]
I have added this in the project structures.
I don't have anything else in the project that is related to spotify.

**Reply by Godwhitelight** [reply]
I made it work now. but i am getting a wird invalid token error.


```
package mamaram.com.dor;


import com.wrapper.spotify.SpotifyApi;
import com.wrapper.spotify.exceptions.SpotifyWebApiException;
import com.wrapper.spotify.model_objects.credentials.ClientCredentials;
import com.wrapper.spotify.model_objects.miscellaneous.CurrentlyPlaying;
import com.wrapper.spotify.requests.authorization.client_credentials.ClientCredentialsRequest;
import com.wrapper.spotify.requests.data.player.GetUsersCurrentlyPlayingTrackRequest;
import org.apache.hc.core5.http.ParseException;

import java.io.IOException;
import java.util.concurrent.CancellationException;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;

public class Main {
    private static final String clientId = "8404e9afd1ef435c990f2841ddad7748";
    private static final String clientSecret = "IAMNOTSTUPID";

    private static final SpotifyApi spotifyApi = new SpotifyApi.Builder()
            .setClientId(clientId)
            .setClientSecret(clientSecret)
            .build();
    private static final ClientCredentialsRequest clientCredentialsRequest = spotifyApi.clientCredentials()
            .build();

    public static void clientCredentials_Sync() {
        try {
            final ClientCredentials clientCredentials = clientCredentialsRequest.execute();

            // Set access token for further "spotifyApi" object usage
            System.out.println(clientCredentials.getAccessToken());
            spotifyApi.setAccessToken(clientCredentials.getAccessToken());
            getUsersCurrentlyPlayingTrack_Sync();
            System.out.println("Expires in: " + clientCredentials.getExpiresIn());
        } catch (IOException | SpotifyWebApiException | ParseException e) {
            System.out.println("Error: " + e.getMessage());
        }
    }
    private static final GetUsersCurrentlyPlayingTrackRequest getUsersCurrentlyPlayingTrackRequest = spotifyApi
            .getUsersCurrentlyPlayingTrack()
//          .market(CountryCode.SE)
//          .additionalTypes("track,episode")
            .build();

    public static void getUsersCurrentlyPlayingTrack_Sync() {
        try {
            final CurrentlyPlaying currentlyPlaying = getUsersCurrentlyPlayingTrackRequest.execute();

            System.out.println("Timestamp: " + currentlyPlaying.getIs_playing());
        } catch (IOException | SpotifyWebApiException | ParseException e) {
            System.out.println("Error: " + e.getMessage());
        }
    }

    public static void getUsersCurrentlyPlayingTrack_Async() {
        try {
            final CompletableFuture<CurrentlyPlaying> currentlyPlayingFuture = getUsersCurrentlyPlayingTrackRequest.executeAsync();

            // Thread free to do other tasks...

            // Example Only. Never block in production code.
            final CurrentlyPlaying currentlyPlaying = currentlyPlayingFuture.join();

            System.out.println("Timestamp: " + currentlyPlaying.getTimestamp());
        } catch (CompletionException e) {
            System.out.println("Error: " + e.getCause().getMessage());
        } catch (CancellationException e) {
            System.out.println("Async operation cancelled.");
        }
    }

    public static void main(String[] args) {
        clientCredentials_Sync();
        //getUsersCurrentlyPlayingTrack_Sync();
    }
}

```

THIS IS THE WHOLE PROJECT.

**Reply by dargmuesli** [reply]
> `IAMNOTSTUPID`

Noone called you stupid. It was recommended to change the secret after it having leaked.

> in the project structures
> `package mamaram.com.dor;`

A Java source file without any ~Gradle~ Maven setup makes it impossible to help. 

> THIS IS THE WHOLE PROJECT.

Capslock is considered screaming here, which lessens people's interest in providing support.

> i am getting a wird invalid token error

See my recommendations above to improve the chance for helpful feedback.

**Reply by Godwhitelight** [reply]
1. I know none called me like that I just changed the secret thing and typed smth random.
2. Wdym? I must have Gradle to use this? I am using the java project and installed a maven repo using the IntelliJ structure library.
3. Noted.
4. the output for the code I sent is VVVVVVV

```
BQDFZ_KU93I-***********************************-J72jq9Cayou9_hhdTdPpWM
Error: Invalid access token
Expires in: 3600
```
the "*" is a char that I replaced for security.

**Reply by dargmuesli** [reply]
Ah sorry, I meant Maven, not Gradle.

Do you know of the [builder pattern](https://en.wikipedia.org/wiki/Builder_pattern)? This library builds upon it.
See, you build the `spotifyApi` object with `.build()`. And **after** that you call `spotifyApi.setAccessToken`. So `spotifyApi` was built without the access token and not built at all after setting the access token. This way the request is made without an access token or - in that sense - with an invalid access token.

**Reply by Godwhitelight** [reply]
ok so now it's fixed. it's not showing that error anymore. but I am getting an Invalid username error
